k8s: Reflector fixes and ListerWatcher test utilities

joamaki opened the pull request:
This adds `pkg/k8s/testutils` package with a decoding utilities and a fake ListerWatcher. These
give a more direct way of writing tests that consume K8s objects than going through the client-go
fake client/tracker. The decoder functions can decode slim and cilium objects from files on disk.

The StateDB k8s reflector tests are reimplemented with the new test utilities and a benchmark is added.

Additionally minor fixes to the reflector to allow using multiple reflectors per table and to transforming
the k8s object into multiple StateDB objects and to be able to merge the existing and new object
(StateDB version is bumped for the Modify() support).

joamaki committed:
k8s: Allow using multiple reflectors per table per module

Add a name parameter in order to have:
- Unique table initializer name
- Unique job name within the module's job group

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
k8s: Add utils for decoding and ListerWatcher faking

Add a canonical decoder for Cilium and Slim objects that we can use in test
across the codebase.

Add FakeListerWatcher that is fed objects directly or via files.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
k8s: Support multiple objects to be returned by reflector transform

Allow the StateDB reflector transform function to return multiple objects,
so that 1-to-many mappings can be supported.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
k8s: Use FakeListerWatcher in reflector test and add benchmark

Use the testutils.FakeListerWatcher to test the reflector. This
simplifies the test as it can be driven more directly and not through
the fake client.

Add a benchmark for the reflector:

goos: linux
goarch: amd64
pkg: github.com/cilium/cilium/pkg/k8s
cpu: 13th Gen Intel(R) Core(TM) i9-13950HX
BenchmarkStateDBReflector-20                  45          25106628 ns/op            796602 objects/sec

Signed-off-by: Jussi Maki <jussi@isovalent.com>

youngnick commented on pkg/k8s/statedb_test.go:
```diff
@@ -30,8 +32,11 @@ func ExampleRegisterReflector() {
-			newTestNodeTable,
+			// Construct the table we're reflecting to.
+			func() statedb.RWTable[*corev1.Node] {
+				// statedb.NewTable(...)
```
Should this be commented? Feels like it either should be uncommented or not present?

joamaki replied:
Good point. I better fill it in for a proper example.

youngnick commented on pkg/k8s/testutils/decoder.go:
```diff
@@ -0,0 +1,39 @@
+	scheme := runtime.NewScheme()
+	slim_fake.AddToScheme(scheme)
+	cilium_v2.AddToScheme(scheme)
+	Decoder = serializer.NewCodecFactory(scheme).UniversalDeserializer()
```
Do you think it would be useful to have a way to add to this scheme later? It will be difficult to use this for Gateway API objects, for example, in the future without a way to add more to the scheme.

I'm not sure how that would interact with other StateDB assumptions though, interested in your thoughts.

joamaki replied:
This isn't related to StateDB in any way (other than that I use it in `pkg/k8s/statedb_test.go` to test it). Definitely the idea is to have everything we use in this scheme. I'll rework it to add everything we use in OSS and also make it extendable from forked codebases.

youngnick reviewed (commented):
Mostly LGTM, just a couple of questions.

joamaki committed:
vendor: Bump StateDB to v0.2.5

Bump StateDB version to pull in the Modify() support.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
k8s: Add Merge option to StateDB reflector

Add the ability to merge the new object with an existing object in the table.
This allows carrying data on the objects that will not be lost when the object
is updated by K8s.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

youngnick reviewed (approved):
LGTM from sig-k8s.

ovidiutirla commented on pkg/k8s/statedb.go:
```diff
@@ -19,44 +19,44 @@ import (
 
 	jobGroup.Add(job.OneShot(
-		fmt.Sprintf("k8s-reflector-[%T]", *new(Obj)),
+		fmt.Sprintf("k8s-reflector-%s", r.ReflectorConfig.Name),
```
nit: should we include the table name too?

joamaki replied:
I think this should be enough. The name does appear qualified with the full module id, so it's easy enough to deduce from context.
EDIT: On second thought, no reason to not be very explicit here. Adding the object type as well.

ovidiutirla reviewed (approved):
Tested the changes part of https://github.com/cilium/cilium/pull/34383. LGTM
